networking/src/Ntp/Client.hs

```diff
@@ -77,280 +80,209 @@ data NtpClient = NtpClient
-    let servers = ntpServers $ ncSettings cli
-    return $ case responsesState of
-        Nothing        -> False
-        Just responses -> length responses >= length servers
+-- Update @'ncStatus'@ according to received responses.
+updateStatus'
+    :: NtpClient
+    -> ([NtpOffset] -> (NtpStatus, (Wlog.Severity, Text)))
+    -> IO ()
+updateStatus' cli fn = do
+    (offset, msg) <- fn <$> readTVarIO (ncState cli)
+    traceWith ntpTrace msg
+    atomically $ writeTVar (ncStatus cli) offset
+
+updateStatus :: NtpClient -> IO ()
+updateStatus cli = updateStatus' cli fn
+    where
+    fn :: [NtpOffset]
+       -> (NtpStatus, (Wlog.Severity, Text))
+    fn [] = ( NtpSyncUnavailable
+            , (Wlog.Warning, "ntp client haven't received any response")
+            )
+    fn offsets =
+        let offset = ntpSelection (ncSettings cli) $ NE.fromList $ offsets
+        in ( NtpDrift offset
+           , (Wlog.Info, sformat ("Evaluated clock offset "%shown%"mcs") offset)
+           )
 
 -- |
--- Low level primitive which sends a request to a single ntp server.
-doSend :: NtpClient -> SockAddr -> IO ()
-doSend cli addr = do
-    sock   <- readTVarIO $ ncSockets cli
-    packet <- encode <$> mkCliNtpPacket
-    handleAny handleE . void $ sendDo addr sock (LBS.toStrict packet)
-  where
-    sendDo a@(SockAddrInet{}) (IPv4Sock sock)      = sendTo' sock a
-    sendDo a@(SockAddrInet{}) (BothSock sock _)    = sendTo' sock a
-    sendDo a@(SockAddrInet6{}) (IPv6Sock sock) = sendTo' sock a
-    sendDo a@(SockAddrInet6{}) (BothSock _ sock)  = sendTo' sock a
-    sendDo a sks                                           =
-        error $ "SockAddr is " <> show a <> ", but sockets: " <> show sks
-    sendTo' sock = flip (sendTo sock)
+-- Every `ntpPollDelay` we send a request to the list of `ntpServers`.  Before
+-- sending a request, we put `NtpSyncPending` to `ncState`.  After sending
+-- all requests we wait until either all servers responded or
+-- `ntpResponseTimeout` passesed.  If at least one server responded
+-- `handleCollectedResponses` will update `ncStatus` in `NtpClient` with a new
+-- drift.
+sendLoop :: NtpClient -> [Addresses] -> IO ()
+sendLoop cli addrs = do
 
-    -- just log; socket closure is handled by receiver
-    handleE =
-        logWarning . sformat ("Failed to send to "%shown%": "%shown) addr
 
--- |
--- Every `ntpPollDelay` send request to the list of `ntpServers`.  Before
--- sending the request, fill `ncState` with `NtpSyncPending`.  After sending
--- requests wait until either all servers respond or `ntpResponseTimeout`
--- passes.  If at least one server responded `handleCollectedResponses` will
--- update `ncStatus` in `NtpClient`.
-startSend :: NtpClient -> [SockAddr] -> IO ()
-startSend cli addrs = do
     let respTimeout = ntpResponseTimeout (ncSettings cli)
-    let poll = ntpPollDelay (ncSettings cli)
-
+    let poll        = ntpPollDelay (ncSettings cli)
+
+    _ <- withAsync
```

Comment:
That looks like you are discarding the result of a computation. Since the last statement in the computation is `cancel a` whose result type is `IO ()`, the above could be written as:
```
    () <- withAsync
```
which is more obviously correct and hence easier for the reviewer.

Reply:
That's true.  PureScript has a class for types which can be ignored without a warning (with an instance for `()`), it would be nice to have it in Haskell.